sql/core/src/main/scala/org/apache/spark/sql/execution/ui/AllExecutionsPage.scala

```diff
@@ -229,73 +406,88 @@ private[ui] abstract class ExecutionTable(
-    showFailedJobs = true) {
 
-  override protected def header: Seq[String] =
-    baseHeader ++ Seq("Running Job IDs", "Succeeded Job IDs", "Failed Job IDs")
-}
+private[ui] class ExecutionTableRowData(
+    val submissionTime: Long,
+    val duration: Long,
+    val executionUIData: SQLExecutionUIData)
+
 
-private[ui] class CompletedExecutionTable(
+private[ui] class ExecutionDataSource(
+    request: HttpServletRequest,
     parent: SQLTab,
+    executionData: Seq[SQLExecutionUIData],
+    basePath: String,
     currentTime: Long,
-    executionUIDatas: Seq[SQLExecutionUIData])
-  extends ExecutionTable(
-    parent,
-    "completed-execution-table",
-    currentTime,
-    executionUIDatas,
-    showRunningJobs = false,
-    showSucceededJobs = true,
-    showFailedJobs = false) {
+    pageSize: Int,
+    sortColumn: String,
+    desc: Boolean) extends PagedDataSource[ExecutionTableRowData](pageSize) {
 
-  override protected def header: Seq[String] = baseHeader ++ Seq("Job IDs")
-}
+  // Convert ExecutionData to ExecutionTableRowData which contains the final contents to show
+  // in the table so that we can avoid creating duplicate contents during sorting the data
+  private val data = executionData.map(executionRow).sorted(ordering(sortColumn, desc))
 
-private[ui] class FailedExecutionTable(
-    parent: SQLTab,
-    currentTime: Long,
-    executionUIDatas: Seq[SQLExecutionUIData])
-  extends ExecutionTable(
-    parent,
-    "failed-execution-table",
-    currentTime,
-    executionUIDatas,
-    showRunningJobs = false,
-    showSucceededJobs = true,
-    showFailedJobs = true) {
+  private var _slicedJobIds: Set[Int] = _
+
+  override def dataSize: Int = data.size
+
+  override def sliceData(from: Int, to: Int): Seq[ExecutionTableRowData] = {
+    val r = data.slice(from, to)
+    _slicedJobIds = r.map(_.executionUIData.executionId.toInt).toSet
+    r
+  }
 
-  override protected def header: Seq[String] =
-    baseHeader ++ Seq("Succeeded Job IDs", "Failed Job IDs")
+  private def executionRow(executionUIData: SQLExecutionUIData): ExecutionTableRowData = {
+    val submissionTime = executionUIData.submissionTime
+    val duration = executionUIData.completionTime.map(_.getTime())
+      .getOrElse(currentTime) - submissionTime
+
+    new ExecutionTableRowData(
+      submissionTime,
+      duration,
+      executionUIData)
+  }
+
+  /**
+    * Return Ordering according to sortColumn and desc
+    */
+  private def ordering(sortColumn: String, desc: Boolean): Ordering[ExecutionTableRowData] = {
+    val ordering: Ordering[ExecutionTableRowData] = sortColumn match {
+      case "ID" => Ordering.by(_.executionUIData.executionId)
+      case "Description" => Ordering.by(_.executionUIData.description)
+      case "Submitted" => Ordering.by(_.executionUIData.submissionTime)
+      case "Duration" => Ordering.by(_.duration)
+      case "Job IDs" | "Succeeded Job IDs" => Ordering by (_.executionUIData.jobs.filter {
```

Comment:
The expression here is still a little complex for my taste, but can't be simplified much. Maybe omit jobId in the case statement, and the second toString below it? the Seq[String] should already have the same ordering.

Reply:
I have modified the code by pre computing the job data based on status and put it in the 'ExecutionTableRowData'.
Now from there, both 'ordering' and 'jobLinks' uses the job data.
Thanks you.